Sources/SwiftSyntax/SyntaxChildren.swift

```diff
@@ -10,305 +10,246 @@
+  /// generated from these.
+  private let rawChildren: PresentRawSyntaxChildren
 
-    public mutating func next() -> Syntax? {
-      return iterator.next()
-    }
-  }
+  /// The parent node of the children. Used to build the `_SyntaxBase` nodes.
+  private let parent: _SyntaxBase
 
-  let node: _SyntaxBase
+  var startIndex: SyntaxChildrenIndex { return rawChildren.startIndex }
+  var endIndex: SyntaxChildrenIndex { return rawChildren.endIndex }
 
-  init(_ node: _SyntaxBase) {
-    self.node = node
+  func index(after index: SyntaxChildrenIndex) -> SyntaxChildrenIndex {
+    return rawChildren.index(after: index)
   }
 
-  public init(_ node: Syntax) {
-    self.node = node.base
+  func index(before index: SyntaxChildrenIndex) -> SyntaxChildrenIndex {
+    return rawChildren.index(before: index)
   }
 
-  public func makeIterator() -> Iterator {
-    return Iterator(node: node)
+  subscript(position: SyntaxChildrenIndex) -> _SyntaxBase {
+    let child = rawChildren[position]
+    let data = SyntaxData(child, parent: parent)
+    return makeSyntax(data)
   }
 
-  public func reversed() -> ReversedSyntaxChildren {
-    return ReversedSyntaxChildren(node)
+  init(_ node: _SyntaxBase) {
+    self.rawChildren = PresentRawSyntaxChildren(node)
+    self.parent = node
   }
 }
 
-/// Reversed Sequence of `SyntaxChildren`.
-public struct ReversedSyntaxChildren: Sequence {
-  public struct Iterator: IteratorProtocol {
-    var iterator: ReversedSyntaxBaseChildren.Iterator
+/// Collection that contains the present child `Syntax` nodes of the given node.
+public struct SyntaxChildren: BidirectionalCollection {
+  public typealias Index = SyntaxChildrenIndex
+  public typealias Element = Syntax
 
-    init(node: _SyntaxBase) {
-      self.iterator = .init(node: node)
-    }
+  /// The underlying collection of `_SyntaxBase` children. `Syntax` nodes are
+  /// generated from these.
+  private let baseChildren: SyntaxBaseChildren
 
-    public mutating func next() -> Syntax? {
-      return iterator.next()
-    }
+  public var startIndex: SyntaxChildrenIndex {
+    return baseChildren.startIndex
+  }
+  public var endIndex: SyntaxChildrenIndex {
+    return baseChildren.endIndex
   }
 
-  let node: _SyntaxBase
-
-  init(_ node: _SyntaxBase) {
-    self.node = node
+  public func index(after index: SyntaxChildrenIndex) -> SyntaxChildrenIndex {
+    return baseChildren.index(after: index)
   }
 
-  public init(_ node: Syntax) {
-    self.node = node.base
+  public func index(before index: SyntaxChildrenIndex) -> SyntaxChildrenIndex {
+    return baseChildren.index(before: index)
   }
 
-  public func makeIterator() -> Iterator {
-    return Iterator(node: node)
+  public subscript(position: SyntaxChildrenIndex) -> Syntax {
+    return baseChildren[position]
   }
 
-  public func reversed() -> SyntaxChildren {
-    return SyntaxChildren(node)
+  public init(_ node: Syntax) {
+    self.baseChildren = SyntaxBaseChildren(node.base)
   }
```

Comment:
You should add a constructor that accepts a `_SyntaxBase`, so that internally we can construct this without doing the unnecessary check of `node.base`.